### ISSUE TITLE
Convenient way to automatically add conversion to custom types?

### ISSUE THREAD

**Issue by vmalloc** [rust-lang-nursery/failure#268]
Sorry for the newbie question, couldn't find any resource or pointer to answer this.

Assuming I have a custom error like the following:
```rust
#[derive(Fail, Debug)]
pub enum MyError {
    #[fail(display = "Custom error 1")]
    CustomError1,

    #[fail(display = "Custom error 2")]
    CustomError2,

}
```
And would like to use it with the following function:
```rust

fn do_something -> Result<(), MyError>
{
    let result: Result<(), std::io::Error> = operation();
    result?
}
```
Any easy/declarative way to say that `io::Error` should map to `CustomErro2`? Do I have to implement std::convert::From myself in such cases?

Thanks in advance

**Comment by KillTheMule** [comment]
Wouldn't https://doc.rust-lang.org/std/result/enum.Result.html#method.map_err suit that?

**Comment by vmalloc** [comment]
@KillTheMule yeah but that would require adding `map_err` in any place that might raise... I was wondering if there is a way to add them automatically through the macro system, so that I wouldn't have to...

**Comment by mitsuhiko** [comment]
@vmalloc the way you *should* be using failure is to not make an error kind per type of source error but by type of failure. So instead of having an `IoError` you have a `CouldNotReadConfig` etc. That's why no such conversion exists.

This is the rule that boats originally came up with and it's what most people do when they use failure.

**Comment by vmalloc** [comment]
I get the general idea behind it, but technically this amounts to loads of `map_err`s throughout the code... I feel like I'm missing something. For example if configuration is stored across several files, every single I/O operation needs to map manually to `CouldNotReadConfig`, which is not very ergnomic....

**Comment by scottjmaddox** [comment]
@vmalloc You might consider merging all lower-level operations into one private function that returns the low-level error, and then having a single `map_err` that maps to your public `Fail` type.
